### DISCUSSION TITLE
Self-Attention-Guidance (SAG) for ComfyUI

### DISCUSSION THREAD

**Discussion by ghost** [comfyanonymous/ComfyUI#2155] [Ideas]
Have been searching for this. I always used FreeU together with SAG in A1111 to eliminate some of the artifacts that SD 1.5 and finetuned models bring. It can bring out some more texture and clean up strange artifacts in the background. Does it exist for ComfyUI?

**Comment by Slickytail** [comment]
Hello, I'm working on implementing this and would appreciate help from anyone. My fork is [here](https://github.com/Slickytail/ComfyUI-Sag).

~I've never used SAG in A1111, so I'm not sure if the results I'm getting are right. Strictly speaking, everything *works* -- there's no crashes etc. However, it gets really ugly if I push the SAG guidance scale past 0.25. The things I'm not sure about:~
 - ~am I pulling the attention map from the right layer?~
 - ~am I correctly computing the estimate x0?~

EDIT: everything works now.
If anyone wants to test this, the SAG node in my fork is `_for_testing/Self-Attention Guidance`.

**Reply by ghost** [reply]
```
Error occurred when executing KSampler:

shape '[1, 17, 12]' is invalid for input of size 165

File "P:\cui\ComfyUI\execution.py", line 153, in recursive_execute
output_data, output_ui = get_output_data(obj, input_data_all)
File "P:\cui\ComfyUI\execution.py", line 83, in get_output_data
return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
File "P:\cui\ComfyUI\execution.py", line 76, in map_node_over_list
results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
File "P:\cui\ComfyUI\nodes.py", line 1299, in sample
return common_ksampler(model, seed, steps, cfg, sampler_name, scheduler, positive, negative, latent_image, denoise=denoise)
File "P:\cui\ComfyUI\nodes.py", line 1269, in common_ksampler
samples = comfy.sample.sample(model, noise, steps, cfg, sampler_name, scheduler, positive, negative, latent_image,
File "P:\cui\ComfyUI\custom_nodes\ComfyUI-Impact-Pack\modules\impact\sample_error_enhancer.py", line 22, in informative_sample
raise e
File "P:\cui\ComfyUI\custom_nodes\ComfyUI-Impact-Pack\modules\impact\sample_error_enhancer.py", line 9, in informative_sample
return original_sample(*args, **kwargs) # This code helps interpret error messages that occur within exceptions but does not have any impact on other operations.
File "P:\cui\ComfyUI\custom_nodes\ComfyUI-AnimateDiff-Evolved\animatediff\sampling.py", line 178, in animatediff_sample
return orig_comfy_sample(model, noise, *args, **kwargs)
File "P:\cui\ComfyUI\comfy\sample.py", line 100, in sample
samples = sampler.sample(noise, positive_copy, negative_copy, cfg=cfg, latent_image=latent_image, start_step=start_step, last_step=last_step, force_full_denoise=force_full_denoise, denoise_mask=noise_mask, sigmas=sigmas, callback=callback, disable_pbar=disable_pbar, seed=seed)
File "P:\cui\ComfyUI\comfy\samplers.py", line 774, in sample
return sample(self.model, noise, positive, negative, cfg, self.device, sampler, sigmas, self.model_options, latent_image=latent_image, denoise_mask=denoise_mask, callback=callback, disable_pbar=disable_pbar, seed=seed)
File "P:\cui\ComfyUI\comfy\samplers.py", line 680, in sample
samples = sampler.sample(model_wrap, sigmas, extra_args, callback, noise, latent_image, denoise_mask, disable_pbar)
File "P:\cui\ComfyUI\comfy\samplers.py", line 619, in sample
samples = self.sampler_function(model_k, noise, sigmas, extra_args=extra_args, callback=k_callback, disable=disable_pbar, **self.extra_options)
File "P:\cui\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
return func(*args, **kwargs)
File "P:\cui\ComfyUI\comfy\k_diffusion\sampling.py", line 207, in sample_dpm_2
denoised = model(x, sigma_hat * s_in, **extra_args)
File "P:\cui\python_embeded\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
return forward_call(*args, **kwargs)
File "P:\cui\ComfyUI\comfy\samplers.py", line 340, in forward
out = self.inner_model(x, sigma, cond=cond, uncond=uncond, cond_scale=cond_scale, model_options=model_options, seed=seed)
File "P:\cui\python_embeded\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
return forward_call(*args, **kwargs)
File "P:\cui\ComfyUI\comfy\samplers.py", line 330, in forward
return self.apply_model(*args, **kwargs)
File "P:\cui\ComfyUI\comfy\samplers.py", line 327, in apply_model
out = sampling_function(self.inner_model, x, timestep, uncond, cond, cond_scale, model_options=model_options, seed=seed)
File "P:\cui\ComfyUI\comfy\samplers.py", line 266, in sampling_function
degraded = create_blur_map(uncond_pred, uncond_attn, sag_sigma, sag_threshold)
File "P:\cui\ComfyUI\comfy\samplers.py", line 291, in create_blur_map
mask.reshape(b, *middle_layer_latent_size)
```

This happens only after I upscale a generated image, turn it into a latent, feed it with SAG into a KSampler for img2img. I attached the workflow. I remember having to use about 0.75 SAG guidance without FreeU and 0.25 - .35 SAG guidance with FreeU activated on A1111.

edit: Quality enhancement definitely works. It is the same error when trying to apply the earlier (better) version of kohya hires fix ( node https://gist.github.com/laksjdjf/487a28ceda7f0853094933d2e138e3c6 )

[SAGerror.json](https://github.com/comfyanonymous/ComfyUI/files/13589105/SAGerror.json)

**Reply by Slickytail** [reply]
Fixed ! Apparently, when using nonstandard"image sizes, the "image" shape in the middle block of the UNet isn't an integer division of the original latent size, so trying to divide the latent dimensions gives the wrong result for the attention map shape. Instead, we can just add a hook to the middle of the UNet to check the data dimensions and smuggle them out.

**Reply by ghost** [reply]
Awesome, thank you so much for your work

**Reply by frankyifei** [reply]
why is it slow down sampling so much, like 40% slower than not using it. Fooocus also has SAG built-in, it does not seem to slow down sampling at all. @Slickytail

**Reply by Slickytail** [reply]
@frankyifei SAG is 1 extra UNet call per diffusion step. Fooocus implements this by skipping the normal conditional step and simply incorporating the SAG blurring into the conditional step (see line 236 of modules/patch.py in fooocus codebase). This is a substantial modification to the recommended cfg+sag formula from the original paper. I have faith in Zhang if he thinks this works better, but in my comfy patch I chose to be more faithful to the paper.

**Reply by rollingcookies** [reply]
Hello, is there any way to modify the node to work in two modes - as it is now and as it is implemented in Fooocus?

**Reply by Slickytail** [reply]
@rollingcookies You would need to make some internal changes to the comfy sampling system. In `comfy.samplers.sampling_function`, you can skip the uncond pass, but not the cond pass. So you'll need to make a few changes there, as well as in `comfy.samplers.calc_cond_uncond_batch` in order to enable that. 

(an easier way would just be to pass through the uncond as the cond, and then make SAG extract the attention map from the cond instead of the uncond, but this might break some other nodes/techniques that rely on modifying only the cond or only the uncond pass).

Then in the sag node, set whatever flag you created to skip the cond pass. Now in the `post_cfg_function`, just change the extra call to `calc_cond_uncond_batch` to pass through the *cond* instead. Finally, change the max scale in the SAG Node, because that becomes your new CFG scale.

**Reply by rollingcookies** [reply]
@Slickytail Thanks for your explanation, I would love to implement all this, but unfortunately I'm not a programmer, just a designer ;(

**Reply by rollingcookies** [reply]
@Slickytail I can help to test this node (and make examples of workflows) if you help write it!
